### DISCUSSION TITLE
Lifecycle functions around the rendering?

### DISCUSSION THREAD

**Discussion by D8H** [4ian/GDevelop#3762] [Enhancement ideas (and their technical discussions)]
## Camera extension conflicts
The screen shake extension conflicts with other extensions that moves the camera like the smooth camera following.
* https://github.com/GDevelopApp/GDevelop-extensions/pull/387

### Current shaking implementation
It seems that the screen shake extension keeps track of of the sum of displacement applied to the camera to put it back where it should be after when the shaking ends.

The issue is that the events or other extensions see the impact of shaking on the camera position. An action to follow an object will greatly reduce the shaking by trying to put back the camera on the object.

### Another broken solution with preEvent
The shaking displacement could be reverted in `preEvent` this works well to avoid conflict with scene events. But as another extension `preEvent` can be called before the shaking one, it would revert the changes done by this other extension.

## Lifecycle functions around the rendering
The only solution I see to solve this conflict is to add new life-cycle functions around the rendering. It will allows to revert the camera position after the rending.

I don't see a lot of other use-case for it, but there could be position rounding on specific object for instance. Maybe it can be a new callback that can only be used with JavaScript.

**Comment by tristanbob** [comment]
I understand the problem, but I don't understand how your suggested solution would work.  Can you explain it more?

One idea to work around this might be for the game developer to add a "Camera is NOT shaking" condition on the "Follow player" action.  However, this could also look bad as the player could move out of the screen during the shake.  

Any other ideas?

**Reply by D8H** [reply]
The life-cycle functions are executed like this:
```
- Behavior pre-events
- Extensions pre-events

- scene events

- Behavior post-events
- Extensions post-events

- Rendering
```

I suggest to add new ones around the rendering:
```
- Behavior pre-events
- Extensions pre-events

- scene events

- Behavior post-events
- Extensions post-events


- Behavior pre-rendering
- Extensions pre-rendering

- Rendering

- Behavior post-rendering
- Extensions post-rendering
```

**Reply by Bouh** [reply]
Two elements at the end are not same as the two first elements? 
- Behavior post-rendering
- Extensions post-rendering

Are equal to:
- Behavior pre-events
- Extensions pre-events

No?

**Reply by D8H** [reply]
It's not the same because there are several behaviors or extensions and their execution order can't be chosen.

A practical example: the shaking extension reverts changes done on the camera to avoid side effects with scene events or other extensions.

With current life-cycle function, the camera following could happen before the revert and thus the character won't be followed.
```
- pre-events
  - Smoothly move the camera on the player
  - Revert the shaking on camera

- scene events

- post-events
  - Apply the shaking on camera

- Rendering
```

With the new life-cycle functions, extensions can choose to do something around events or around rendering. As it's a loop, it's, indeed, as breaking pre-events and post-events in 2 priorities.
```
- pre-events
  - Smoothly move the camera on the player

- scene events

- post-events


- pre-rendering
  - Apply the shaking on camera

- Rendering

- post-rendering
  - Revert the shaking on camera
```
